sklearn/metrics/_regression.py

```diff
@@ -868,12 +891,14 @@ def _assemble_r2_explained_variance(
         nonzero_numerator = numerator != 0
         # Default = Zero Numerator = perfect predictions. Set to 1.0
         # (note: even if denominator is zero, thus avoiding NaN scores)
-        output_scores = np.ones([n_outputs])
+        output_scores = xp.ones([n_outputs], device=device_, dtype=dtype)
         # Non-zero Numerator and Non-zero Denominator: use the formula
         valid_score = nonzero_denominator & nonzero_numerator
-        output_scores[valid_score] = 1 - (
+
+        output_scores[valid_score] = xp.ones(1, device=device_, dtype=dtype) - (
```

Comment:
It is not super clear to me why we need to create the array here. I would expect the same on the above line:

```python
output_scores = 1 - (numerator / denominator)
```

but checking a bit it seems that when dealing with tensor, broadcasting is working the same way.

So just curious, what is the reason to have to create this array of `ones` here.

Reply:
I agree, the `xp.ones(1, device=device_, dtype=dtype)` does not seem necessary. `1` should work.

Reply:
Similar discussion than https://github.com/scikit-learn/scikit-learn/pull/27904#discussion_r1482911579